### PR TITLE
terminal: enable link detectors in detached terminals

### LINKED ISSUE #188055: Support link detection in test results terminal
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version: VSCode 1.80.1, 1.81.0-insider; plugin playwright test for VSCode 1.0.14
- OS Version: Windows 10

Steps to Reproduce:
- Run playwright TS test throwing an error. 
- Open test results viewer in vs code (see picture)
- Try Ctrl+Click on call stack item for the error 

**Expected**
We jump to source code line

**Actual**
Nothing happens.

It stopped to work ~ a week ago, I guess. But used to work in  vs code insider build. But today is stopped to work in the insider build too. (monday, 17.07.2023)
Doesn't reproduce in VSCode  1.79.2.

Скриншот 17-07-2023 105240

### PR DESCRIPTION
**Pull request by connor4312**
Initially the direction was to make `TerminalLinkManager` more general
and use it explicitly in detached terminals.

But I ended up getting nudged in the direction of making terminalContribs
work with detached terminals instead. All the link detection logic is
in terminalContrib which is a layer that can't be imported from the main
terminal code. And it also depends on the raw xterm instance, which we
don't want to expose to callers, so I wouldn't want to have the caller
manually do something like new `TerminalLinkManager(deatchedXterm.raw, ...)`

So that made me think I should either make a new contrib system for
detached terminals, or allow existing contribs to signal that they
could run for detached terminals too. In the PR, I did the latter.

Fixes https://github.com/microsoft/vscode/issues/188055

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/browser/terminal.ts [hunk 1]
```diff
@@ -146,8 +146,26 @@ export interface IDetachedXTermOptions {
 	colorProvider: IXtermColorProvider;
 	capabilities?: ITerminalCapabilityStore;
 	readonly?: boolean;
+	processInfo: ITerminalProcessInfo;
 }
 
+export interface IDetachedTerminalInstance extends IDisposable {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminal.ts:152]: A comment explaining that it emulates an `ITerminalInstance` in order to enable more features would be handy in the future.

**Review by Tyriar — COMMENTED** [review]
Overall idea looks good :+1:

File: src/vs/workbench/contrib/terminal/browser/terminalExtensions.ts [hunk 2]
```diff
@@ -5,19 +5,22 @@
...
+export type DetachedTerminalContributionCtor = IConstructorSignature<ITerminalContribution, [IDetachedTerminalInstance, ITerminalProcessInfo, TerminalWidgetManager]>;
 
-export interface ITerminalContributionDescription {
-	readonly id: string;
-	readonly ctor: TerminalContributionCtor;
-}
+export type ITerminalContributionDescription = { readonly id: string } & (
+	| { readonly canRunInDetachedTerminals: false; readonly ctor: TerminalContributionCtor | DetachedTerminalContributionCtor }
+	| { readonly canRunInDetachedTerminals: true; readonly ctor: DetachedTerminalContributionCtor }
+);
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalExtensions.ts:18]: I was expecting this?

```ts
	| { readonly canRunInDetachedTerminals: false; readonly ctor: TerminalContributionCtor }
	| { readonly canRunInDetachedTerminals: true; readonly ctor: TerminalContributionCtor 
 | DetachedTerminalContributionCtor }
```

File: src/vs/workbench/contrib/terminal/browser/terminalService.ts [hunk 3]
```diff
@@ -1018,7 +1019,7 @@ export class TerminalService extends Disposable implements ITerminalService {
 		return this._createTerminal(shellLaunchConfig, location, options);
 	}
 
-	async createDetachedXterm(options: IDetachedXTermOptions): Promise<IDetachedXtermTerminal> {
+	async createDetachedTerminal(options: IDetachedXTermOptions): Promise<IDetachedTerminalInstance> {
 		const ctor = await TerminalInstance.getXtermConstructor(this._keybindingService, this._contextKeyService);
 		const instance = this._instantiationService.createInstance(
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalService.ts:1024]: Let's rename `instance` to `xterm` or similar

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/terminal/browser/terminalExtensions.ts (hunk 2 above)
Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/terminalExtensions.ts:18]: I think I understand what you're going for, but it's pretty confusing to look at and took me a while to understand:

```ts
canRunInDetachedTerminals: false; readonly ctor: ... | DetachedTerminalContributionCtor
```

How about this?

```ts
export type TerminalContributionCtor = IConstructorSignature<ITerminalContribution, [ITerminalInstance, ITerminalProcessManager, TerminalWidgetManager]>;
export type DetachedCompatibleTerminalContributionCtor = IConstructorSignature<ITerminalContribution, [IDetachedTerminalInstance, ITerminalProcessInfo, TerminalWidgetManager]>;

export type ITerminalContributionDescription = { readonly id: string } & (
	| { readonly canRunInDetachedTerminals: false; readonly ctor: TerminalContributionCtor }
	| { readonly canRunInDetachedTerminals: true; readonly ctor: DetachedCompatibleTerminalContributionCtor }
```

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/terminal/browser/terminalExtensions.ts (hunk 2 above)
Comment (connor4312) [reply src/vs/workbench/contrib/terminal/browser/terminalExtensions.ts:18]: yep, that works

**Comment by Tyriar** [comment]
Can we also make the `DetachedTerminalContributionCtor` -> `DetachedCompatibleTerminalContributionCtor` change? That makes it much clearer that it will work on regular terminals as well

**Review by Tyriar — APPROVED** [review]

**Pull request merged**